cmd/influxd/upgrade/security_test.go

```diff
@@ -0,0 +1,248 @@
+						"hits": influxql.WritePrivilege,
+					},
+				},
+				{
+					Name:  "boss@hits.org", // special name
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"hits": influxql.AllPrivileges,
+					},
+				},
+				{
+					Name:  "viewer",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"water": influxql.ReadPrivilege,
+						"air": influxql.ReadPrivilege,
+					},
+				},
+			},
+			db2ids: map[string][]string{
+				"water": {"33f9d67bc9cbc5b7", "33f9d67bc9cbc5b8", "33f9d67bc9cbc5b9"},
+				"air": {"43f9d67bc9cbc5b7", "43f9d67bc9cbc5b8", "43f9d67bc9cbc5b9"},
+				"hits": {"53f9d67bc9cbc5b7"},
+			},
+			want: testScriptShOrdinary,
+		},
+		{
+			name: "missing buckets",
+			users: []meta.UserInfo{
+				{
+					Name:  "weatherman",
+					Admin: false,
+					Privileges: map[string]influxql.Privilege{
+						"water": influxql.AllPrivileges,
+						"air": influxql.AllPrivileges,
+					},
+				},
+			},
+			db2ids: nil,
+			wantErr: errors.New("upgrade: there were errors/warnings, please fix them and run the command again"),
+		},
+		{
+			name: "no users",
+			users: []meta.UserInfo{},
+			want: "",
+		},
+	}
+
+	var shSuffix string
+	if runtime.GOOS == "win" {
+		shSuffix = ".cmd"
+	} else {
+		shSuffix = ".sh"
+	}
+	shSuffix = ".cmd"
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) { // better do not run in parallel
+			// mock v1 meta
+			v1 := &influxDBv1{
+				meta: &meta.Client{},
+			}
+			data := &meta.Data{
+				Users: tc.users,
+			}
+			// inject users into mock v1 meta client
+			f := reflect.ValueOf(v1.meta).Elem().Field(4)
+			f = reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr())).Elem()
+			f.Set(reflect.ValueOf(data))
+
+			// script output file
+			tmpfile, err := ioutil.TempFile("", "upgrade-security-*" + shSuffix)
+			if err != nil {
+				t.Fatal(err)
+			}
+			tmpfile.Close()
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer os.Remove(tmpfile.Name())
+
+			// options passed on cmdline
+			targetOptions := optionsV2{
+				userName: "admin",
+				orgName: "demo",
+				token: "ABC007==",
+				securityScriptPath: tmpfile.Name(),
+			}
+
+			// command execution
+			err = generateSecurityScript(v1, targetOptions, tc.db2ids, zap.NewNop())
+			if err != nil {
+				if tc.wantErr != nil {
+					if diff := cmp.Diff(tc.wantErr.Error(), err.Error()); diff != "" {
+						t.Fatal(diff)
+					}
+				} else {
+					t.Fatal(err)
+				}
+			}
+
+			// validate result by comparing arrays of non-empty lines of wanted vs actual output
+			parse := func(content string) []string {
+				var lines []string
+				scanner := bufio.NewScanner(strings.NewReader(content))
+				for scanner.Scan() {
+					line := strings.TrimSpace(scanner.Text())
+					if line != "" {
+						// In `influx create auth` command line, read/write bucket args may be in different order,
+						// because user privileges is a map. So reorder them alphabetically for easier comparison.
```

Comment:
This is not true, as the user privileges is keyed in a map, but stored in a slice, which has a consistent ordering.

Can this code be simplified based on that information?

Reply:
The comment is not good, I've modified it. The order of bucket IDs is random on the generated command line because order of `database` names in iteration over `meta.UserInfo.Privileges` map in random (https://github.com/influxdata/influxdb/blob/af20f784b2765604fc10eb1012b93405b7087138/cmd/influxd/upgrade/security.go#L96). So buckets permissions has to be sorted to get the same line every test run.

Reply:
The overarching goal is the script output should be deterministic.

Ultimately that will require sorting any source data to ensure that property is maintained.

Reply:
Now the list authorizations is based on sorted database names list (lines 95-101 in https://github.com/influxdata/influxdb/blob/f6cf893ba6dcd6e940ec6e6c8865b82a480b39df/cmd/influxd/upgrade/security.go#L95).